File: baekjoon/sw 역량테스트 기출/경사로.py
def check_up(y,x):
    if x-L==0:
        return True
    elif x-L>0 and matrix[y][x-L-1]<=matrix[y][x-L]:
        return True
    else:
        return False

def check_down(y,x,v):
    for j in range(x+1,x+L):
        if x+L<=N and matrix[y][j] == v:
            continue
        else:
            return False
    else:
        if x+L == N:
            return N-1
        elif x+L < N and matrix[y][x+L-1] >= matrix[y][x+L]:
            return x+L-1
        else:
            return False

def check():
    global road
    for r in range(N):
        c = cnt = 1
        val = matrix[r][0]
        while c != N:
            if matrix[r][c] == val:
                cnt += 1
            elif val - matrix[r][c] == -1 and c == N - 1:
                if cnt >= L and check_up(r, c):  # 마지막 열일때, cnt를 확인안해서 틀렸음
                    pass
                else:
                    break
            elif val - matrix[r][c] == -1 and cnt >= L:
                if check_up(r, c):
                    cnt, val = 1, matrix[r][c]
                else:
                    break
            elif val - matrix[r][c] == 1:
                c = check_down(r, c, matrix[r][c])
                if c:
                    cnt, val = 0, matrix[r][c]
                else:
                    break
            else:
                break
            c += 1
        if c == N:
            road += 1

N,L = map(int,input().split())
matrix = [list(map(int,input().split())) for _ in range(N)]
matrix_rotate = [[0]*N for _ in range(N)]
road = 0

check()
for j in range(N):
    for i in range(N):
        matrix_rotate[j][i]=matrix[i][j]
for i in range(N):
    for j in range(N):
        matrix[i][j] = matrix_rotate[i][j]
check()

print(road)

# def check_up(y,x):
#     if x-L==0:
#         return True
#     elif x-L>0 and matrix[y][x-L-1]<=matrix[y][x-L]:
#         return True
#     else:
#         return False
#
# def check_down(y,x,v):
#     for j in range(x+1,x+L):
#         if x+L<=N and matrix[y][j] == v:
#             continue
#         else:
#             return False
#     else:
#         if x+L == N:
#             return N-1
#         elif x+L < N and matrix[y][x+L-1] >= matrix[y][x+L]:
#             return x+L-1
#         else:
#             return False
#
# N,L = map(int,input().split())
# matrix = [list(map(int,input().split())) for _ in range(N)]
# matrix_rotate = [[0]*N for _ in range(N)]
# road = 0
#
# for r in range(N):
#     c = cnt = 1
#     val = matrix[r][0]
#     while c!=N:
#         if matrix[r][c] == val:
#             cnt += 1
#         elif val-matrix[r][c]==-1 and c==N-1:
#             if cnt>=L and check_up(r,c): #마지막 열일때, cnt를 확인안해서 틀렸음
#                 pass
#             else:
#                 break
#         elif val-matrix[r][c]==-1 and cnt>=L:
#             if check_up(r,c):
#                 cnt,val = 1,matrix[r][c]
#             else:
#                 break
#         elif val-matrix[r][c]==1:
#             c = check_down(r,c,matrix[r][c])
#             if c:
#                 cnt,val = 0, matrix[r][c]
#             else:
#                 break
#         else:
#             break
#         c += 1
#     if c == N:
#         # print('{} right'.format(r))
#         road += 1
#
# for j in range(N):
#     for i in range(N):
#         matrix_rotate[j][i]=matrix[i][j]
#
# for i in range(N):
#     for j in range(N):
#         matrix[i][j] = matrix_rotate[i][j]
#
# for r in range(N):
#     c = cnt =1
#     val = matrix[r][0]
#     while c!=N:
#         if matrix[r][c] == val:
#             cnt += 1
#         elif val-matrix[r][c]==-1 and c==N-1:
#             if cnt>=L and check_up(r,c):
#                 pass
#             else:
#                 break
#         elif val-matrix[r][c]==-1 and cnt>=L:
#             if check_up(r,c):
#                 cnt,val = 1,matrix[r][c]
#             else:
#                 break
#         elif val-matrix[r][c]==1:
#             c = check_down(r,c,matrix[r][c])
#             if c:
#                 cnt,val = 0, matrix[r][c]
#             else:
#                 break
#         else:
#             break
#         c += 1
#     if c == N:
#         # print('{} down'.format(r))
#         road += 1
#
# print(road)

File: baekjoon/c/Project2/Project2/경사로.c
#if 0
#include <stdio.h>
#include <stdbool.h>

bool checkUp(int y, int *x, int l, int s[][100]);
int checkDown(int y, int *x, int v, int l, int s[][100], int N);

int main(void) {
	freopen("°æ»ç·Î.txt", "r", stdin);
	setbuf(stdout, NULL);

	int matrix[100][100] = { 0 };
	int matrix_rotate[100][100] = { 0 };
	int i, j, N, L, r, c, cnt,val;
	int road = 0;
	int *p;
	int *pc;
	p = matrix;
	pc = &c;

	scanf("%d %d", &N, &L);
	for (i = 0; i < N; i++) {
		for (j = 0; j < N; j++) {
			scanf("%d", &matrix[i][j]);
		}
	}

	for (r = 0; r < N; r++) {
		c = cnt = 1;
		val = matrix[r][0];
		while (c != N) {
			if (matrix[r][c] == val) {
				cnt++;
			}
			else if (val - matrix[r][c] == -1 && N - 1) {
				if (cnt >= L && checkUp(r, c, L, p)) {
					c += 1;
					continue;
				}
				else {
					break;
				}
			}
			else if (val - matrix[r][c] == -1 && cnt >= L) {
				if (checkUp(r, c, L,p)) {
					cnt = 1;
					val = matrix[r][c];
				}
				else {
					break;
				}
			}
			else if (val - matrix[r][c] == 1) {
				printf("%d", c);
				c = checkDown(r, pc, matrix[r][c], L, p, N);
				printf("%d\n", c);
				if (c!=-1) {
					cnt = 0;
					val = matrix[r][c];
				}
				else {
					break;
				}
			}
			else {
				break;
			}
			c++;
		}
		if (c == N) {
			printf("%d right\n", r);
			road++;
		}
	}

	for (j = 0; j < N; j++) {
		for (i = 0; i < N; i++) {
			matrix_rotate[j][i] = matrix[i][j];
		}
	}
	for (i = 0; i < N; i++) {
		for (j = 0; j < N; j++) {
			matrix[i][j] = matrix_rotate[i][j];
		}
	}

	for (r = 0; r < N; r++) {
		c = cnt = 1;
		val = matrix[r][0];
		while (c != N) {
			if (matrix[r][c] == val) {
				cnt++;
			}
			else if (val - matrix[r][c] == -1 && N - 1) {
				if (cnt >= L && checkUp(r, c, L, p)) {
					c += 1;
					continue;
				}
				else {
					break;
				}
			}
			else if (val - matrix[r][c] == -1 && cnt >= L) {
				if (checkUp(r, c, L, p)) {
					cnt = 1;
					val = matrix[r][c];
				}
				else {
					break;
				}
			}
			else if (val - matrix[r][c] == 1) {
				c = checkDown(r, pc, matrix[r][c], L, p, N);
				if (c!=-1) {
					cnt = 0;
					val = matrix[r][c];
				}
				else {
					break;
				}
			}
			else {
				break;
			}
			c++;
		}
		if (c == N) {
			printf("%d down\n", r);
			road++;
		}
	}

	printf("%d", road);
	return 0;
}

bool checkUp(int y, int *x, int l, int s[][100]) {
	if (*x - l == 0)
		return true;
	else if (*x - l > 0 && s[y][*x - l - 1] <= s[y][*x - l])
		return true;
	return false;
}

int checkDown(int y, int *x, int v,int l,int s[][100], int N) {
	for (int j = x + 1; j < *x + l;j++) {
		if (*x + l <= N && s[y][j] == v) {
			continue;
		}
		else {
			return -1;
		}
	}
	if (*x + l == N){
		return N - 1;
	}
	else if (*x + l < N && s[y][*x + l - 1] >= s[y][*x + l]) {
		return x + l - 1;
	}
	else {
		return -1;
	}
}
#endif